src/platforms/common/stm32/traceswoasync.c

```diff
@@ -106,35 +136,90 @@ void traceswo_setspeed(uint32_t baudrate)
 	usart_enable_rx_dma(SWO_UART);
 }
 
+#if defined(DMA_STREAM0)
 void SWO_DMA_ISR(void)
 {
-	if (DMA_ISR(SWO_DMA_BUS) & DMA_ISR_HTIF(SWO_DMA_CHAN)) {
-		DMA_IFCR(SWO_DMA_BUS) |= DMA_ISR_HTIF(SWO_DMA_CHAN);
+	if (dma_get_interrupt_flag(SWO_DMA_BUS, SWO_DMA_CHAN, DMA_HTIF)) {
+		dma_clear_interrupt_flags(SWO_DMA_BUS, SWO_DMA_CHAN, DMA_HTIF);
 		memcpy(&trace_rx_buf[write_index * TRACE_ENDPOINT_SIZE], pingpong_buf, TRACE_ENDPOINT_SIZE);
 	}
-	if (DMA_ISR(SWO_DMA_BUS) & DMA_ISR_TCIF(SWO_DMA_CHAN)) {
-		DMA_IFCR(SWO_DMA_BUS) |= DMA_ISR_TCIF(SWO_DMA_CHAN);
+	if (dma_get_interrupt_flag(SWO_DMA_BUS, SWO_DMA_CHAN, DMA_TCIF)) {
+		dma_clear_interrupt_flags(SWO_DMA_BUS, SWO_DMA_CHAN, DMA_TCIF);
 		memcpy(
 			&trace_rx_buf[write_index * TRACE_ENDPOINT_SIZE], &pingpong_buf[TRACE_ENDPOINT_SIZE], TRACE_ENDPOINT_SIZE);
 	}
 	write_index = (write_index + 1U) % NUM_TRACE_PACKETS;
 	trace_buf_drain(usbdev, TRACE_ENDPOINT | USB_REQ_TYPE_IN);
 }
+#else
+void SWO_DMA_ISR(void)
+{
+	if (dma_get_interrupt_flag(SWO_DMA_BUS, SWO_DMA_CHAN, DMA_HTIF)) {
+		dma_clear_interrupt_flags(SWO_DMA_BUS, SWO_DMA_CHAN, DMA_HTIF);
+		memcpy(&trace_rx_buf[write_index * TRACE_ENDPOINT_SIZE], pingpong_buf, TRACE_ENDPOINT_SIZE);
+	}
+	if (dma_get_interrupt_flag(SWO_DMA_BUS, SWO_DMA_CHAN, DMA_TCIF)) {
+		dma_clear_interrupt_flags(SWO_DMA_BUS, SWO_DMA_CHAN, DMA_TCIF);
+		memcpy(
+			&trace_rx_buf[write_index * TRACE_ENDPOINT_SIZE], &pingpong_buf[TRACE_ENDPOINT_SIZE], TRACE_ENDPOINT_SIZE);
+	}
+	write_index = (write_index + 1U) % NUM_TRACE_PACKETS;
+	trace_buf_drain(usbdev, TRACE_ENDPOINT | USB_REQ_TYPE_IN);
+}
+#endif
 
 void traceswo_init(uint32_t baudrate, uint32_t swo_chan_bitmask)
 {
+#ifdef TRACESWOASYNC_ALLOCS
+	if (trace_rx_buf == NULL) {
+		uint8_t *newbuf = memalign(TRACE_ENDPOINT_SIZE, NUM_TRACE_PACKETS * TRACE_ENDPOINT_SIZE);
```

Comment:
`memalign()` is not a C standard library function (it's not defined by the standard, that is), and we're not sure what the impact will be on newlib under a very constrained memory environment. How important is it that the allocation be aligned to the trace endpoint size? We don't recall seeing any source buffer alignment requirements, so unless there are any, we strongly suggest using `malloc()` instead here.

Reply:
This buffer used to have static storage duration (and linkage), but despite it being uint8_t[] BFD linker guarantees some alignment, likely 32 bit (check this). Now that it's allocated dynamically, either malloc or memalign should ensure 32-bit alignment for 1) DMA to work properly. Although USART Rx streams are configured with `DMA_MSIZE_8BIT`. 2) This will be memcpy'ed in IRQ context to USB buffers, either 16-bit PMA or OTG FIFO, so I want 32-bit alignment for an efficient memcpy4 (`__aeabi_memcpy4`?) to pump data faster than a byte-wise memcpy1. Alignment for **64 bytes** is not supported by anything, but I wanted to see if it's honored at all by newlib.

Reply:
Dropped memalign.
I have two suggestions, 1) `return false` to command.c (requires changing prototypes) on allocation failure, else user sees "Trace enabled" which is a lie. 2) Attempt to allocate for a smaller buffer size (down to 2x64 bytes) and optionally report this via `gdb_voutf()`. Which of those are necessary -- will only be visible in testing.

Reply:
Definitely thing that, if you're happy to do a prototype change, having the commands able to report failures has the potential to be extremely useful and good for reducing what they have to do - "failed? just return false". Quite happy for that to be done.